### PR TITLE
Supporting smart crop

### PR DESCRIPTION
**Pull request by Dynom**
**Status: Ready for merge**

For this PoC I've abused `bimg.Gravity*`'s gravity constants to determine wether or not we should apply the SmartCrop algorithm. It works, [feel free to test it](https://hub.docker.com/r/dynom/imaginary/tags/). However I don't like this approach for two reasons:
* It would introduce a `Gravity` type in `bimg` that isn't a valid `Gravity` type for `libvips`
* It indirectly creates a change in `bimg` that is only needed for `Imaginary` (Which would be an incorrect responsibility relation)

I think there are a couple of options to solve the design issue:
- Generalising `parseGravity()` so that it'll return an `int`. This makes it more generic and it's easy to implement. However we'll lose some type coherence. It also means we'll need to introduce another `magic` value. It's possible and easy to implement, but I'm not too fond of it.
- Introduce boilerplate on several places, to work around it. I don't like this approach as it clutters the code. It is easy to implement however.
- Introduce a completely independent parameter (`?smartcrop=true`). I don't like this approach, since the instructions seem mutually-exclusive (`?gravity=north&smartcrop=true` makes no sense).
- Introducing specific types in Imaginary to solve the responsibility problem. This means duplication of the `Gravity*` constants, but it's better for the (de-)coupling between `bimg` and Imaginary. Most significantly the relation bimg->Imaginary. It's possible and doesn't change the `bimg` implementation. I like this approach.
- Introducing the GravitySmart constant in `bimg` and let `bimg` know how to deal with it. Overall this might be the best option, but also requires the most work. I like this approach best.

Other Changes:
* libvips is now downloaded directly from Github. This gives us access to the more recent versions.
* I've changed some `if` conditions to a [map lookup](https://github.com/h2non/imaginary/pull/136/files#diff-698bf4e6eb6b63e163486214fb780438L165). This is mostly taste, but I (personally) prefer the readability. If we agree on that I would like to change this on a few other places as well. I also prefer to make these changes in a different PR all together.

### CODE REVIEW

**Comment by Dynom** [comment]
After discussing it with @h2non we've picked the option "Introducing the GravitySmart constant in `bimg` and let `bimg` know how to deal with it".

I'll prepare a PR for https://github.com/h2non/bimg and update this PR to reflect those changes.

**Comment by Dynom** [comment]
Before the merge I'll simplify the commits history, I'll make it two commits, one for the feature and one for some of the cosmetic changes.

Or you can leave it like this and flatten the history with a squash merge.

**Comment by Dynom** [comment]
FWIW: The build fails because bimg is pinned on `.v1` and my commit isn't part of a `1.x` tag yet. Locally it builds just fine

**Comment by h2non** [comment]
I have just pushed the new tag version for `bimg`. It should work now.

File: README.md [hunk 1]
```diff
@@ -271,9 +271,9 @@ Usage:
   imaginary -enable-url-source -allowed-origins http://localhost,http://server.com
   imaginary -enable-url-source -enable-auth-forwarding
   imaginary -enable-url-source -authorization "Basic AwDJdL2DbwrD=="
-	imaginary -enable-placeholder
-	imaginery -enable-url-source -placeholder ./placeholder.jpg
-	imaginary -h | -help
+  imaginary -enable-placeholder
+  imaginery -enable-url-source -placeholder ./placeholder.jpg
```
Comment (h2non) [inline README.md]: Typo. This should be `imaginary`. Probably my fault a long time ago.

File: image.go [hunk 2]
```diff
@@ -40,7 +40,7 @@ func Info(buf []byte, o ImageOptions) (Image, error) {
 
 	meta, err := bimg.Metadata(buf)
 	if err != nil {
-		return image, NewError("Cannot retrieve image medatata: %s"+err.Error(), BadRequest)
+		return image, NewError("Cannot retrieve image metadata: %s"+err.Error(), BadRequest)
```
Comment (h2non) [inline image.go]: Thanks for that typo fix :)

File: imaginary.go [hunk 3]
```diff
@@ -60,9 +60,9 @@ Usage:
   imaginary -enable-url-source -allowed-origins http://localhost,http://server.com
   imaginary -enable-url-source -enable-auth-forwarding
   imaginary -enable-url-source -authorization "Basic AwDJdL2DbwrD=="
-	imaginary -enable-placeholder
-	imaginery -enable-url-source -placeholder ./placeholder.jpg
-	imaginary -h | -help
+  imaginary -enable-placeholder
+  imaginery -enable-url-source -placeholder ./placeholder.jpg
```
Comment (h2non) [inline imaginary.go]: Same here.

**Comment by h2non** [comment]
LGTM. Once the build passes, I will merge this and release a new imaginary version. I'm even considering bumping it to be a major `v1` version. Thoughts?

File: imaginary.go (hunk 3 above)
Comment (Dynom) [inline imaginary.go]: Ah, good catch, missed that.. will fix it right away

**Comment by Dynom** [comment]
For some reason gopkg still references bimg's v1.0.8, not v1.0.9. It's probably cache that needs expiring. I've created an issue with gopkg (https://github.com/niemeyer/gopkg/issues/56), just in case.


Another reason to start with versioning (-;

**Comment by Dynom** [comment]
There is a typo in the tag name, it's now named `v.1.0.9` (extra dot) instead of `v1.0.9` so gopkg doesn't pick it up.

**Comment by Dynom** [comment]
I'm impartial to the versioning. It's a nice feature and it might be nice to hit a v1. Imaginary isn't a library kind of project, so versioning is less strict I suppose.

So sure, bump it :confetti_ball:

**Comment by h2non** [comment]
Yeah, in this case it's for maturity level, mostly. However, versioning here is also important, even if it's not a programmatic oriented package. People using docker can rely on that.

File: imaginary.go [hunk 4]
```diff
@@ -241,7 +241,7 @@ func parseOrigins(origins string) []*url.URL {
 func memoryRelease(interval int) {
 	ticker := time.NewTicker(time.Duration(interval) * time.Second)
 	go func() {
-		for _ = range ticker.C {
+		for range ticker.C {
```
Comment (tomas-fp) [inline imaginary.go]: This is not a valid syntax in Go 1.3 or lower versions. I'm open to officially Go 1.5+ from now. 
Can you update `.travis.yml`?

See: https://travis-ci.org/h2non/imaginary/jobs/236177548#L3727-L3728

Comment (Dynom) [inline imaginary.go]: Do we still plan to support 3 year old versions.. ?

Comment (tomas-fp) [inline imaginary.go]: No. As I mentioned, the idea here is to support Go 1.5+.

Comment (Dynom) [inline imaginary.go]: In that case we might have even more merit to bump the version to the big 1.0. I've added the two other versions as an `allow_failures` directive. I could also just remove 'm from the list, not sure what's desired at this point.

**Comment by Dynom** [comment]
The build passes, ready for a merge and version bump @h2non ?

**Pull request merged**